### PR TITLE
merge fedavg.py with fedavgwithEarlyStopping

### PR DESCRIPTION
**Pull request by chesterxgchen**
Fixes # .

### Description

A few sentences describing the changes proposed in this pull request.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/app_common/workflows/fedavg.py [hunk 1]
```diff
@@ -47,14 +146,239 @@ def run(self) -> None:
...
+
+        target_metric, _, op_fn = self.stop_condition
+        curr_target_metric = curr_metrics.get(target_metric, None)
+        if curr_target_metric is None:
+            return False
+
+        if self.best_target_metric_value is None or op_fn(curr_target_metric, self.best_target_metric_value):
+            if self.patience and self.best_target_metric_value == curr_target_metric:
+                self.num_fl_rounds_without_improvement += 1
+                return False
```
[the remark this answers is not in the document]
Comment (chesterxgchen) [reply nvflare/app_common/workflows/fedavg.py:331]: this is not a logic error. I don't see error here.

File: nvflare/app_common/workflows/fedavg.py [hunk 2]
```diff
@@ -12,31 +12,130 @@
...
+        # Load initial model - prefer initial_model if provided, else use persistor
+        if self.initial_model is not None:
+            if isinstance(self.initial_model, FLModel):
+                model = self.initial_model
+            else:
+                # Assume dict of params
+                model = FLModel(params=self.initial_model)
+            self.info("Using provided initial_model")
+        else:
+            model = self.load_model()
```
Comment (YuanTingHsieh) [inline nvflare/app_common/workflows/fedavg.py:137]: maybe should update the load_model method in "BaseModelController" class if we want to support this.

File: nvflare/app_opt/pt/fedavg_early_stopping.py [hunk 3]
```diff
@@ -12,193 +12,136 @@
...
-        """
-        super().__init__(*args, **kwargs)
-        self.patience = patience
-        self.task_to_optimize = task_to_optimize
-        self.num_fl_rounds_without_improvement: int = 0
-        self.stop_cond = stop_cond
-        if self.stop_cond:
-            self.stop_condition = parse_compare_criteria(stop_cond)
+        # Convert PyTorch model to dict if needed
+        if initial_model is not None and hasattr(initial_model, "state_dict"):
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/pt/fedavg_early_stopping.py]: instead of hasattr, we also need to check if it is a callable,
should we just do if isinstance(initial_model, torch.nn.Module):

File: nvflare/app_opt/pt/fedavg_early_stopping.py [hunk 4]
```diff
@@ -12,193 +12,136 @@
...
+        elif isinstance(initial_model, dict):
+            initial_model_params = initial_model
+        elif isinstance(initial_model, FLModel):
+            initial_model_params = initial_model
         else:
-            self.stop_condition = None
-        self.save_filename = save_filename
-        self.initial_model: FLModel = initial_model
-        self.best_target_metric_value: Any = None
+            initial_model_params = initial_model
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/pt/fedavg_early_stopping.py]: should we raise exception if the type is not those 3?

File: nvflare/app_opt/pt/fedavg.py [hunk 5]
```diff
@@ -12,193 +12,136 @@
...
-            clients = self.sample_clients(self.num_clients)
+        """Run FedAvg workflow with PyTorch tensor serialization support."""
+        # Register TensorDecomposer for FOBS serialization of PyTorch tensors
+        fobs.register(TensorDecomposer)
+        super().run()
 
-            results: List[FLModel] = self.send_model_and_wait(
-                task_name=self.task_to_optimize, targets=clients, data=model
-            )
+    def save_model_file(self, model: FLModel, filepath: str) -> None:
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/pt/fedavg.py:110]: why we cant use FOBs directly? @nvidianz

File: nvflare/app_common/workflows/fedavg.py [hunk 6]
```diff
@@ -12,31 +12,130 @@
...
 
+    def __init__(
+        self,
+        *args,
+        initial_model: Optional[Union[Dict, FLModel]] = None,
+        save_filename: Optional[str] = "FL_global_model.pt",
+        aggregator: Optional[ModelAggregator] = None,
+        stop_cond: Optional[str] = None,
+        patience: Optional[int] = None,
+        task_to_optimize: Optional[str] = "train",
```
Comment (YuanTingHsieh) [inline nvflare/app_common/workflows/fedavg.py]: this is a weird arg name IMO, but not sure what's the best.
@holgerroth @ZiyueXu77 any ideas

File: nvflare/app_opt/pt/fedavg_early_stopping.py [hunk 7]
```diff
@@ -12,193 +12,136 @@
...
 from nvflare.app_common.abstract.fl_model import FLModel
-from nvflare.app_common.utils.math_utils import parse_compare_criteria
-from nvflare.app_common.workflows.base_fedavg import BaseFedAvg
+from nvflare.app_common.workflows.fedavg import FedAvg
 from nvflare.app_opt.pt.decomposers import TensorDecomposer
 from nvflare.fuel.utils import fobs
 
 
-class PTFedAvgEarlyStopping(BaseFedAvg):
+class PTFedAvgEarlyStopping(FedAvg):
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/pt/fedavg_early_stopping.py]: this should probably just be called PTFedAvg

File: nvflare/app_opt/tf/recipes/fedavg.py [hunk 8]
```diff
@@ -55,11 +54,6 @@ class FedAvgRecipe(UnifiedFedAvgRecipe):
         params_transfer_type (str): How to transfer the parameters. FULL means the whole model parameters are sent.
             DIFF means that only the difference is sent. Defaults to TransferType.FULL.
         model_persistor: Custom model persistor. If None, TFModelPersistor will be used.
-        analytics_receiver: Component for receiving analytics data (e.g., TBAnalyticsReceiver for TensorBoard).
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/tf/recipes/fedavg.py:58]: this removal makes sense, we just promote using the "add_experiment_tracking" method, @nvkevlu @holgerroth @ZiyueXu77 we need to make sure the examples dont use this pattern.

File: nvflare/app_opt/pt/recipes/fedavg.py [hunk 9]
```diff
@@ -46,79 +45,48 @@ class FedAvgRecipe(UnifiedFedAvgRecipe):
...
-            train_script="client.py",
-            train_args="--epochs 5 --batch_size 32"
-        )
-        ```
+        per_site_config: Per-site configuration for the federated learning job.
+        key_metric: Metric used to determine if the model is globally best. Defaults to "accuracy".
+        stop_cond: Early stopping condition based on metric. String literal in the format of
+            '<key> <op> <value>' (e.g. "accuracy >= 80"). If None, early stopping is disabled.
+        patience: Number of rounds with no improvement after which FL will be stopped.
+        task_to_optimize: Task name for training. Defaults to "train".
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/pt/recipes/fedavg.py]: maybe for recipes we dont need to expose this "task" name

File: nvflare/recipe/fedavg.py [hunk 10]
```diff
@@ -39,17 +36,25 @@ class _FedAvgValidator(BaseModel):
...
-    analytics_receiver: Any
-    per_site_config: Optional[dict[str, dict]] = None
-    key_metric: str
+    model_persistor: Optional[ModelPersistor] = None
+    per_site_config: Optional[Dict[str, Dict]] = None
+    key_metric: str = "accuracy"
+    # New FedAvg features
+    stop_cond: Optional[str] = None
+    patience: Optional[int] = None
+    task_to_optimize: str = "train"
```
Comment (YuanTingHsieh) [inline nvflare/recipe/fedavg.py]: maybe for recipes we dont need to expose this "task" name

File: nvflare/recipe/fedavg.py [hunk 11]
```diff
@@ -268,6 +292,67 @@ def __init__(
...
+
+        # Try PyTorch nn.Module
+        if hasattr(self.initial_model, "state_dict"):
+            return self.initial_model.state_dict()
+
+        # Try TensorFlow/Keras model
+        if hasattr(self.initial_model, "get_weights"):
+            weights = self.initial_model.get_weights()
+            # Convert to dict with layer indices as keys
+            return {f"layer_{i}": w for i, w in enumerate(weights)}
```
Comment (YuanTingHsieh) [inline nvflare/recipe/fedavg.py]: I think these should be in child recipes not move up here, these might change as those lib version changed, and I feel its more suitable there.

the check is better to be "if isinstance(initial_model, torch.nn.Module):" as well

Comment (YuanTingHsieh) [reply nvflare/recipe/fedavg.py]: since child logic already handled by "_setup_model_and_persistor" maybe we just remove these parts here

File: nvflare/recipe/fedavg.py [hunk 12]
```diff
@@ -268,6 +292,67 @@ def __init__(
...
+
+        # Unknown type - try to use as-is
+        return self.initial_model
```
Comment (YuanTingHsieh) [inline nvflare/recipe/fedavg.py]: should we do this or raise exception?

**Review by YuanTingHsieh — COMMENTED** [review]
suggest some changes

File: nvflare/app_common/aggregators/weighted_aggregation_helper.py [hunk 13]
```diff
@@ -70,11 +106,32 @@ def add(self, data, weight, contributor_name, contribution_round):
...
+        Uses in-place division to minimize memory copies.
+        Returns the aggregated dict and resets internal state.
+        """
         with self.lock:
-            aggregated_dict = {k: v * (1.0 / self.counts[k]) for k, v in self.total.items()}
-            self.reset_stats()
-            return aggregated_dict
+            # In-place division where possible
+            for k, v in self.total.items():
+                scale = 1.0 / self.counts[k]
```
[the remark this answers is not in the document]
Comment (chesterxgchen) [reply nvflare/app_common/aggregators/weighted_aggregation_helper.py]: we are not chaning this file, this is out dated

File: nvflare/app_opt/pt/fedavg.py [hunk 14]
```diff
@@ -12,193 +12,136 @@
...
-            raise RuntimeError(f"stop criteria key '{key}' doesn't exists in metrics")
-
-        if op_fn(value, target):
-            self.info(f"Early stop condition satisfied: {self.stop_condition}")
-            return True
+        # Save FLModel metadata (metrics, params_type, etc.) separately
+        params = model.params
+        model.params = {}  # Temporarily remove params to save only metadata
+        fobs.dumpf(model, f"{filepath}.metadata")
+        model.params = params  # Restore params
```
[the remark this answers is not in the document]
Comment (chesterxgchen) [reply nvflare/app_opt/pt/fedavg.py:126]: this is FLModel not tensor model. THis is ok

File: nvflare/app_common/workflows/fedavg.py (hunk 2 above)
Comment (chesterxgchen) [reply nvflare/app_common/workflows/fedavg.py:137]: good point, that was my intention, as the original fedavgwithearly stopping was doing that. 

I will check on this

File: nvflare/app_opt/pt/fedavg_early_stopping.py (hunk 3 above)
Comment (chesterxgchen) [reply nvflare/app_opt/pt/fedavg_early_stopping.py]: good one

File: nvflare/app_common/workflows/fedavg.py (hunk 6 above)
Comment (chesterxgchen) [reply nvflare/app_common/workflows/fedavg.py]: This was added by the community contributor ( Forgot which company),  as my original early stopping did not has task. The Client API has different task names. so the contributor add this.

File: nvflare/app_opt/pt/recipes/fedavg.py (hunk 9 above)
Comment (chesterxgchen) [reply nvflare/app_opt/pt/recipes/fedavg.py]: The question is how does the user specify different task, maybe we remove task name, but hard code the task name for each recipe

File: nvflare/app_opt/pt/fedavg_early_stopping.py (hunk 3 above)
Comment (chesterxgchen) [reply nvflare/app_opt/pt/fedavg_early_stopping.py]: addressed

File: nvflare/app_common/workflows/fedavg.py (hunk 2 above)
Comment (chesterxgchen) [reply nvflare/app_common/workflows/fedavg.py:137]: changed

File: nvflare/app_opt/pt/fedavg.py (hunk 5 above)
Comment (chesterxgchen) [reply nvflare/app_opt/pt/fedavg.py:110]: why do we want to use FOBS here ?  I saw the Tensorflow example is using FOBS the model file name is something .fob, looks really strange for the model.

File: nvflare/app_common/workflows/fedavg.py (hunk 6 above)
Comment (chesterxgchen) [reply nvflare/app_common/workflows/fedavg.py]: I removed from recipe but keep it at FedAvg Controller and renamed to task_name

File: nvflare/app_opt/pt/recipes/fedavg.py (hunk 9 above)
Comment (chesterxgchen) [reply nvflare/app_opt/pt/recipes/fedavg.py]: changed

File: nvflare/recipe/fedavg.py (hunk 10 above)
Comment (chesterxgchen) [reply nvflare/recipe/fedavg.py]: changed

File: nvflare/recipe/fedavg.py (hunk 12 above)
Comment (chesterxgchen) [reply nvflare/recipe/fedavg.py]: changed

File: nvflare/recipe/fedavg.py (hunk 11 above)
Comment (chesterxgchen) [reply nvflare/recipe/fedavg.py]: changed

File: nvflare/app_opt/pt/fedavg_early_stopping.py (hunk 7 above)
Comment (chesterxgchen) [reply nvflare/app_opt/pt/fedavg_early_stopping.py]: changed

File: nvflare/app_opt/pt/fedavg.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+            filepath (str): path to save the model
+        """
+        # Save parameters with torch.save
+        torch.save(model.params, filepath)
+
+        # Save FLModel metadata (metrics, params_type, etc.) separately
+        params = model.params
+        model.params = {}  # Temporarily remove params to save only metadata
+        fobs.dumpf(model, f"{filepath}.metadata")
+        model.params = params  # Restore params
```
[the remark this answers is not in the document]
Comment (chesterxgchen) [reply nvflare/app_opt/pt/fedavg.py:126]: not an issue

File: nvflare/app_opt/tf/recipes/fedavg.py [hunk 16]
```diff
@@ -158,11 +122,34 @@ def __init__(
...
+
+        # Try TensorFlow/Keras model
+        if hasattr(self.initial_model, "get_weights"):
+            weights = self.initial_model.get_weights()
+            # Convert to dict with layer indices as keys
+            return {f"layer_{i}": w for i, w in enumerate(weights)}
+
+        raise TypeError(
+            f"initial_model must be a dict, tf.keras.Model, or None. " f"Got {type(self.initial_model).__name__}."
+        )
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/tf/recipes/fedavg.py]: if initial model is not None, then _setup_model_and_persistor already initialize the persistor, so these will never be executed.

File: nvflare/app_opt/pt/recipes/fedavg.py [hunk 17]
```diff
@@ -161,9 +133,38 @@ def __init__(
...
+
+        # Try PyTorch nn.Module
+        from nvflare.fuel.utils.import_utils import optional_import
+
+        nn, torch_ok = optional_import(module="torch.nn")
+        if torch_ok and isinstance(self.initial_model, nn.Module):
+            return self.initial_model.state_dict()
+
+        raise TypeError(
+            f"initial_model must be a dict, torch.nn.Module, or None. " f"Got {type(self.initial_model).__name__}."
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/pt/recipes/fedavg.py]: if initial model is not None, then _setup_model_and_persistor already initialize the persistor, so these will never be executed.

File: nvflare/app_opt/tf/recipes/fedavg.py (hunk 16 above)
Comment (chesterxgchen) [reply nvflare/app_opt/tf/recipes/fedavg.py]: changed

File: nvflare/app_opt/pt/recipes/fedavg.py (hunk 17 above)
Comment (chesterxgchen) [reply nvflare/app_opt/pt/recipes/fedavg.py]: changed

**Review by holgerroth — APPROVED** [review]
Nice new functionality!

File: nvflare/recipe/fedavg.py [hunk 18]
```diff
@@ -268,6 +289,59 @@ def __init__(
...
+        else:
+            # It's a Shareable-based Aggregator - can't use directly with FedAvg
+            # Log a warning and fall back to built-in aggregation
+            import logging
+
+            logging.getLogger(__name__).warning(
+                f"Provided aggregator {type(self.aggregator).__name__} is not a ModelAggregator. "
+                "Using built-in weighted averaging instead. For custom aggregation with FedAvg, "
+                "please use a ModelAggregator subclass (e.g., from model_aggregator.py)."
+            )
```
[the remark this answers is not in the document]
Comment (chesterxgchen) [reply nvflare/recipe/fedavg.py:342]: no. This is just information
